imports/plugins/core/orders/server/no-meteor/mutations/cancelOrderItem.js

```diff
@@ -0,0 +1,185 @@
+  const { Orders } = collections;
+
+  // First verify that this order actually exists
+  const order = await Orders.findOne({ _id: orderId });
+  if (!order) throw new ReactionError("not-found", "Order not found");
+
+  // Allow cancel if the account that placed the order is attempting to cancel
+  // or if the account has "orders" permission. When called internally by another
+  // plugin, context.isInternalCall can be set to `true` to disable this check.
+  if (
+    !isInternalCall &&
+    (!accountId || accountId !== order.accountId) &&
+    !userHasPermission(["orders"], order.shopId)
+  ) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  // Is the account calling this mutation also the account that placed the order?
+  // We need this check in a couple places below, so we'll get it here.
+  const accountIsOrderer = (order.accountId && accountId === order.accountId);
+
+  // The orderer may only cancel while the order status is still "new"
+  if (accountIsOrderer && !orderStatusesThatOrdererCanCancel.includes(order.workflow.status)) {
+    throw new ReactionError("invalid", `Order status (${order.workflow.status}) is not one of: ${orderStatusesThatOrdererCanCancel.join(", ")}`);
+  }
+
+  // Find and cancel the item
+  let foundItem = false;
+  const updatedGroups = order.shipping.map((group) => {
+    let itemToAdd;
+    const updatedItems = group.items.map((item) => {
+      if (item._id !== itemId) return item;
+      foundItem = true;
+
+      // The orderer may only cancel while the order item status is still "new"
+      if (accountIsOrderer && !itemStatusesThatOrdererCanCancel.includes(item.workflow.status)) {
+        throw new ReactionError("invalid", `Item status (${item.workflow.status}) is not one of: ${itemStatusesThatOrdererCanCancel.join(", ")}`);
+      }
+
+      // If they are requesting to cancel fewer than the total quantity of items
+      // that were ordered, we'll create a new item here with the remaining quantity.
+      // It will have the same status as this item has before we cancel it.
+      // Later, after we exit this loop, we'll append it to the `group.items` list.
+      if (item.quantity > cancelQuantity) {
+        itemToAdd = {
+          ...item,
+          _id: Random.id(),
+          quantity: item.quantity - cancelQuantity
+        };
+      } else if (item.quantity < cancelQuantity) {
+        throw new ReactionError("invalid-param", "cancelQuantity may not be greater than item quantity");
+      }
+
+      // If we make it this far, then we've found the item that they want to cancel.
+      // We set the status and the cancel reason if one was provided.
+      // This will also decrement the quantity to match the quantity that is being
+      // canceled, which will be offset by pushing `itemToAdd` into the array later.
+      return {
+        ...item,
+        cancelReason: reason,
```

Comment:
Not sure if this is a big deal, as it doesn't seem to do anything aside from add a new item to the array, but I can run the same mutation multiple times and the workflow updates each time:

<img width="781" alt="robo_3t_-_1_1" src="https://user-images.githubusercontent.com/4482263/53509606-f7398180-3a70-11e9-89ae-c042e80147e9.png">

Reply:
It's probably nicer if we avoid doing an update when it's already canceled. I'll figure out a fix.

Reply:
I pushed a fix. It will still allow the cancelation, so `cancelReason` for example would change if called a second time with a different reason, but it won't change the status or add to workflow array.